src/spatio_temporal_voxel_layer.cpp

```diff
@@ -662,24 +662,28 @@ void SpatioTemporalVoxelLayer::updateCosts( \
 }
 
 /*****************************************************************************/
-void SpatioTemporalVoxelLayer::UpdateROSCostmap(double* min_x, double* min_y, \
-                                                double* max_x, double* max_y)
+void SpatioTemporalVoxelLayer::UpdateROSCostmap( \
+                        double* min_x, double* min_y, \
+                        double* max_x, double* max_y, \
+                        std::unordered_set<volume_grid::occupany_cell>& updated_cells)
 /*****************************************************************************/
 {
   // grabs map of occupied cells from grid and adds to costmap_
   Costmap2D::resetMaps();
 
-  std::unordered_map<volume_grid::occupany_cell, uint>::iterator it;
-  for (it = _voxel_grid->GetFlattenedCostmap()->begin();
-       it != _voxel_grid->GetFlattenedCostmap()->end(); ++it)
+  auto cost_map = _voxel_grid->GetFlattenedCostmap();
```

Comment:
Revert this to iterators and specific calls, I don't understand why you changed this, it includes an entire copy

Reply:
I call the same function as you, `SpatioTemporalVoxelGrid::GetFlattenedCostmap`, which returns a pointer, not an entire copy. If anything, shouldn't my version be faster than yours, since I'm only calling that function once while you call it twice?

Then, I get the iterator I need by calling `unordered_map::find`, which runs in O(1) time.

Looping `updated_cells` is based on references, so I'm not making any copies either.

Reply:
O(1) doesn't mean it isn't slow, it just means its constant-slow, revert please. This is a huge amount of data you're trying to find and its going to ding performance. You'd rather loop through a vector of `M` values and do a thing than search `N` values, where`N` and `M` are similiarly sized due to processor caching.

Reply:
> you'd rather loop through a vector of M values and do a thing than search N values

There is no "searching" of any kind, you are not looping the map and checking every single element for its key.

`unordered_map` is an hash table. `unordered_map::find` simply computes the hash of the given key and returns the corresponding value.

Reply:
I don't think you understand. 

Doing O(1) N times can be slower than O(N) 1 time, you're equating orders of scale with the literal compute time. `find()` might be O(1) but its not a **fast** O(1), iterating through the entire container is nearly continuous memory so we can rely on caching the search can't. 

You can test this yourself using a chrono timer, I can almost guarantee this would be non-trivially slower

Reply:
Ah, I see. Thanks for explaining!